File: android/app/src/main/java/com/francescooddo/buddygrammar/core/KeyboardInteractionRouter.kt
package com.francescooddo.buddygrammar.core

import kotlin.math.abs
import kotlin.math.hypot
import kotlin.math.roundToInt

data class InteractionPoint(
    val x: Double,
    val y: Double,
) {
    fun distanceTo(other: InteractionPoint): Double = hypot(x - other.x, y - other.y)

    companion object {
        val Zero = InteractionPoint(0.0, 0.0)
    }
}

sealed interface InteractionTarget {
    data class Key(
        val literal: String,
        val alternates: List<String> = emptyList(),
        val allowsSwipe: Boolean = true,
    ) : InteractionTarget

    data object Space : InteractionTarget
    data object Delete : InteractionTarget
}

enum class InteractionFeedback {
    KEY,
    SELECTION,
}

data class InteractionDeadline(
    val token: Int,
    val dueTimeSeconds: Double,
    val kind: Kind,
) {
    enum class Kind {
        LONG_PRESS,
        CURSOR_ACTIVATION,
        DELETE_REPEAT,
    }
}

sealed interface InteractionInput {
    data class Press(
        val target: InteractionTarget,
        val point: InteractionPoint,
        val timeSeconds: Double,
    ) : InteractionInput

    data class Move(
        val point: InteractionPoint,
        val timeSeconds: Double,
    ) : InteractionInput

    data class Release(
        val point: InteractionPoint,
        val timeSeconds: Double,
    ) : InteractionInput

    data class Deadline(val deadline: InteractionDeadline) : InteractionInput
    data object Cancel : InteractionInput
}

sealed interface InteractionEffect {
    data class Pressed(val target: InteractionTarget?) : InteractionEffect
    data class Preview(val text: String?) : InteractionEffect
    data class Schedule(val deadline: InteractionDeadline) : InteractionEffect
    data class ShowAlternates(
        val alternates: List<String>,
        val selectedIndex: Int,
    ) : InteractionEffect

    data object HideAlternates : InteractionEffect
    data class CommitText(val text: String) : InteractionEffect
    data object DeleteBackward : InteractionEffect
    data object DeleteWord : InteractionEffect
    data class MoveCursor(val characterDelta: Int) : InteractionEffect
    data class SwipeBegan(val point: InteractionPoint) : InteractionEffect
    data class SwipeMoved(val point: InteractionPoint) : InteractionEffect
    data class SwipeEnded(val point: InteractionPoint) : InteractionEffect
    data class Feedback(val kind: InteractionFeedback) : InteractionEffect
}

/**
 * Platform-neutral pointer interaction state machine.
 *
 * Scheduling and editor mutation stay in the native adapter. Deadlines carry a
 * generation token, so releasing, cancelling, or starting another key makes an
 * outstanding callback harmless.
 */
class KeyboardInteractionRouter(
    private val configuration: Configuration = KeyboardCatalog.gestures.routerConfiguration(),
) {
    data class Configuration(
        val longPressDelaySeconds: Double = 0.35,
        val swipeDistance: Double = 24.0,
        val alternateStep: Double = 26.0,
        val cursorActivationDelaySeconds: Double = 0.18,
        val cursorActivationDistance: Double = 12.0,
        val cursorStep: Double = 12.0,
        val deleteRepeatDelaySeconds: Double = 0.36,
        val deleteRepeatIntervalSeconds: Double = 0.07,
        val minimumDeleteRepeatIntervalSeconds: Double = 0.07,
    )

    private sealed interface ActiveState {
        data class Key(
            val token: Int,
            val literal: String,
            val alternates: List<String>,
            val origin: InteractionPoint,
            val allowsSwipe: Boolean,
            val isSwiping: Boolean = false,
            val alternatesVisible: Boolean = false,
            val alternateIndex: Int = 0,
        ) : ActiveState

        data class Space(
            val token: Int,
            val origin: InteractionPoint,
            val cursorMode: Boolean = false,
            val emittedCursorSteps: Int = 0,
        ) : ActiveState

        data class Delete(
            val token: Int,
            val repeatCount: Int = 0,
        ) : ActiveState
    }

    private var nextToken = 0
    private var active: ActiveState? = null

    fun handle(input: InteractionInput): List<InteractionEffect> = when (input) {
        is InteractionInput.Press -> begin(input)
        is InteractionInput.Move -> move(input.point, input.timeSeconds)
        is InteractionInput.Release -> release(input.point)
        is InteractionInput.Deadline -> fire(input.deadline)
        InteractionInput.Cancel -> cancel()
    }

    private fun begin(input: InteractionInput.Press): List<InteractionEffect> {
        nextToken += 1
        val token = nextToken
        return when (val target = input.target) {
            is InteractionTarget.Key -> {
                active = ActiveState.Key(
                    token = token,
                    literal = target.literal,
                    alternates = target.alternates,
                    origin = input.point,
                    allowsSwipe = target.allowsSwipe,
                )
                buildList {
                    add(InteractionEffect.Pressed(target))
                    add(InteractionEffect.Preview(target.literal))
                    if (target.alternates.isNotEmpty()) {
                        add(
                            InteractionEffect.Schedule(
                                InteractionDeadline(
                                    token = token,
                                    dueTimeSeconds = input.timeSeconds +
                                        configuration.longPressDelaySeconds.coerceAtLeast(0.0),
                                    kind = InteractionDeadline.Kind.LONG_PRESS,
                                ),
                            ),
                        )
                    }
                    add(InteractionEffect.Feedback(InteractionFeedback.KEY))
                }
            }

            InteractionTarget.Space -> {
                active = ActiveState.Space(
                    token = token,
                    origin = input.point,
                )
                listOf(
                    InteractionEffect.Pressed(InteractionTarget.Space),
                    InteractionEffect.Schedule(
                        InteractionDeadline(
                            token = token,
                            dueTimeSeconds = input.timeSeconds +
                                configuration.cursorActivationDelaySeconds.coerceAtLeast(0.0),
                            kind = InteractionDeadline.Kind.CURSOR_ACTIVATION,
                        ),
                    ),
                    InteractionEffect.Feedback(InteractionFeedback.KEY),
                )
            }

            InteractionTarget.Delete -> {
                active = ActiveState.Delete(token = token)
                listOf(
                    InteractionEffect.Pressed(InteractionTarget.Delete),
                    InteractionEffect.DeleteBackward,
                    InteractionEffect.Feedback(InteractionFeedback.KEY),
                    InteractionEffect.Schedule(
                        InteractionDeadline(
                            token = token,
                            dueTimeSeconds = input.timeSeconds +
                                configuration.deleteRepeatDelaySeconds.coerceAtLeast(0.0),
                            kind = InteractionDeadline.Kind.DELETE_REPEAT,
                        ),
                    ),
                )
            }
        }
    }

    private fun move(
        point: InteractionPoint,
        _timeSeconds: Double,
    ): List<InteractionEffect> = when (val state = active) {
        is ActiveState.Key -> when {
            state.alternatesVisible -> {
                val rawIndex = ((point.x - state.origin.x) /
                    configuration.alternateStep.coerceAtLeast(1.0)).roundToInt()
                val index = rawIndex.coerceIn(0, state.alternates.lastIndex)
                if (index == state.alternateIndex) {
                    emptyList()
                } else {
                    active = state.copy(alternateIndex = index)
                    listOf(InteractionEffect.ShowAlternates(state.alternates, index))
                }
            }

            state.isSwiping -> listOf(InteractionEffect.SwipeMoved(point))
            state.allowsSwipe &&
                state.origin.distanceTo(point) >= configuration.swipeDistance.coerceAtLeast(1.0) -> {
                active = state.copy(isSwiping = true)
                listOf(
                    InteractionEffect.Preview(null),
                    InteractionEffect.SwipeBegan(state.origin),
                    InteractionEffect.SwipeMoved(point),
                )
            }

            else -> emptyList()
        }

        is ActiveState.Space -> {
            val horizontalDistance = point.x - state.origin.x
            if (!state.cursorMode &&
                abs(horizontalDistance) <
                configuration.cursorActivationDistance.coerceAtLeast(1.0)
            ) {
                emptyList()
            } else {
                val effects = mutableListOf<InteractionEffect>()
                if (!state.cursorMode) {
                    effects += InteractionEffect.Preview(null)
                    effects += InteractionEffect.Feedback(InteractionFeedback.SELECTION)
                }
                val totalSteps = (horizontalDistance /
                    configuration.cursorStep.coerceAtLeast(1.0)).toInt()
                val delta = totalSteps - state.emittedCursorSteps
                if (delta != 0) effects += InteractionEffect.MoveCursor(delta)
                active = state.copy(
                    cursorMode = true,
                    emittedCursorSteps = totalSteps,
                )
                effects
            }
        }

        is ActiveState.Delete, null -> emptyList()
    }

    private fun release(point: InteractionPoint): List<InteractionEffect> {
        val released = active
        active = null
        return when (released) {
            is ActiveState.Key -> when {
                released.isSwiping -> listOf(
                    InteractionEffect.SwipeEnded(point),
                    InteractionEffect.Pressed(null),
                )

                released.alternatesVisible -> listOf(
                    InteractionEffect.CommitText(released.alternates[released.alternateIndex]),
                    InteractionEffect.HideAlternates,
                    InteractionEffect.Preview(null),
                    InteractionEffect.Pressed(null),
                )

                else -> listOf(
                    InteractionEffect.CommitText(released.literal),
                    InteractionEffect.Preview(null),
                    InteractionEffect.Pressed(null),
                )
            }

            is ActiveState.Space -> if (released.cursorMode) {
                listOf(InteractionEffect.Pressed(null))
            } else {
                listOf(
                    InteractionEffect.CommitText(" "),
                    InteractionEffect.Pressed(null),
                )
            }

            is ActiveState.Delete -> listOf(InteractionEffect.Pressed(null))
            null -> emptyList()
        }
    }

    private fun fire(deadline: InteractionDeadline): List<InteractionEffect> =
        when (val state = active) {
            is ActiveState.Key -> if (
                deadline.kind == InteractionDeadline.Kind.LONG_PRESS &&
                deadline.token == state.token &&
                !state.isSwiping &&
                !state.alternatesVisible &&
                state.alternates.isNotEmpty()
            ) {
                active = state.copy(alternatesVisible = true, alternateIndex = 0)
                listOf(
                    InteractionEffect.ShowAlternates(state.alternates, 0),
                    InteractionEffect.Feedback(InteractionFeedback.SELECTION),
                )
            } else {
                emptyList()
            }

            is ActiveState.Delete -> if (
                deadline.kind == InteractionDeadline.Kind.DELETE_REPEAT &&
                deadline.token == state.token
            ) {
                val effect = InteractionEffect.DeleteBackward
                val updated = state.copy(repeatCount = state.repeatCount + 1)
                active = updated
                val baseInterval = configuration.deleteRepeatIntervalSeconds.coerceAtLeast(0.01)
                val minimumInterval = configuration.minimumDeleteRepeatIntervalSeconds
                    .coerceIn(0.01, baseInterval)
                val acceleration = (updated.repeatCount * 0.01)
                    .coerceAtMost(baseInterval - minimumInterval)
                listOf(
                    effect,
                    InteractionEffect.Feedback(InteractionFeedback.KEY),
                    InteractionEffect.Schedule(
                        InteractionDeadline(
                            token = updated.token,
                            dueTimeSeconds = deadline.dueTimeSeconds + baseInterval - acceleration,
                            kind = InteractionDeadline.Kind.DELETE_REPEAT,
                        ),
                    ),
                )
            } else {
                emptyList()
            }

            is ActiveState.Space -> if (
                deadline.kind == InteractionDeadline.Kind.CURSOR_ACTIVATION &&
                deadline.token == state.token &&
                !state.cursorMode
            ) {
                active = state.copy(cursorMode = true)
                listOf(
                    InteractionEffect.Preview(null),
                    InteractionEffect.Feedback(InteractionFeedback.SELECTION),
                )
            } else {
                emptyList()
            }

            null -> emptyList()
        }

    private fun cancel(): List<InteractionEffect> {
        val cancelled = active ?: return emptyList()
        active = null
        return when (cancelled) {
            is ActiveState.Key -> if (cancelled.alternatesVisible) {
                listOf(
                    InteractionEffect.HideAlternates,
                    InteractionEffect.Preview(null),
                    InteractionEffect.Pressed(null),
                )
            } else {
                listOf(InteractionEffect.Preview(null), InteractionEffect.Pressed(null))
            }

            is ActiveState.Space, is ActiveState.Delete ->
                listOf(InteractionEffect.Pressed(null))
        }
    }
}

File: BuddyGrammarKit/Sources/BuddyGrammarKit/KeyboardInteractionRouter.swift
import Foundation

/// Platform-neutral coordinates used by the keyboard interaction state machine.
/// The rendering adapter decides whether a unit represents points or pixels.
public struct InteractionPoint: Equatable, Sendable {
    public let x: Double
    public let y: Double

    public init(x: Double, y: Double) {
        self.x = x
        self.y = y
    }

    public static let zero = InteractionPoint(x: 0, y: 0)

    fileprivate func distance(to other: InteractionPoint) -> Double {
        hypot(x - other.x, y - other.y)
    }
}

public enum KeyboardInteractionTarget: Equatable, Sendable {
    case key(String, alternates: [String], allowsSwipe: Bool = true)
    case space
    case delete
}

public enum KeyboardInteractionFeedback: Equatable, Sendable {
    case key
    case selection
}

public struct KeyboardInteractionDeadline: Equatable, Sendable {
    public enum Kind: Equatable, Sendable {
        case longPress
        case cursorActivation
        case deleteRepeat
    }

    public let token: Int
    public let dueTime: TimeInterval
    public let kind: Kind

    public init(token: Int, dueTime: TimeInterval, kind: Kind) {
        self.token = token
        self.dueTime = dueTime
        self.kind = kind
    }
}

public enum KeyboardInteractionInput: Equatable, Sendable {
    case press(
        target: KeyboardInteractionTarget,
        at: InteractionPoint,
        time: TimeInterval
    )
    case move(to: InteractionPoint, time: TimeInterval)
    case release(at: InteractionPoint, time: TimeInterval)
    case deadline(KeyboardInteractionDeadline)
    case cancel
}

public enum KeyboardInteractionEffect: Equatable, Sendable {
    case pressed(KeyboardInteractionTarget?)
    case preview(String?)
    case schedule(KeyboardInteractionDeadline)
    case showAlternates([String], selectedIndex: Int)
    case hideAlternates
    case commitText(String)
    case deleteBackward
    case deleteWord
    case moveCursor(Int)
    case swipeBegan(InteractionPoint)
    case swipeMoved(InteractionPoint)
    case swipeEnded(InteractionPoint)
    case feedback(KeyboardInteractionFeedback)
}

/// Serializes key preview, long-press, swipe, cursor, and delete-repeat rules.
///
/// The router is deliberately pure with respect to the editor and clock. The
/// platform schedules emitted deadlines and feeds them back; stale deadlines
/// are ignored by token. This keeps pointer arbitration identical on iOS and
/// Android without making literal insertion depend on asynchronous work.
public struct KeyboardInteractionRouter: Sendable {
    public struct Configuration: Equatable, Sendable {
        public var longPressDelay: TimeInterval
        public var swipeDistance: Double
        public var alternateStep: Double
        public var cursorActivationDistance: Double
        public var cursorActivationDelay: TimeInterval
        public var cursorStep: Double
        public var deleteRepeatDelay: TimeInterval
        public var deleteRepeatInterval: TimeInterval
        public var minimumDeleteRepeatInterval: TimeInterval
        public var wordDeleteAfterRepeats: Int

        public init(
            longPressDelay: TimeInterval = 0.35,
            swipeDistance: Double = 24,
            alternateStep: Double = 26,
            cursorActivationDistance: Double = 12,
            cursorActivationDelay: TimeInterval = 0.18,
            cursorStep: Double = 14,
            deleteRepeatDelay: TimeInterval = 0.45,
            deleteRepeatInterval: TimeInterval = 0.11,
            minimumDeleteRepeatInterval: TimeInterval = 0.045,
            wordDeleteAfterRepeats: Int = 6
        ) {
            self.longPressDelay = max(0, longPressDelay)
            self.swipeDistance = max(1, swipeDistance)
            self.alternateStep = max(1, alternateStep)
            self.cursorActivationDistance = max(1, cursorActivationDistance)
            self.cursorActivationDelay = max(0, cursorActivationDelay)
            self.cursorStep = max(1, cursorStep)
            self.deleteRepeatDelay = max(0, deleteRepeatDelay)
            self.deleteRepeatInterval = max(0.01, deleteRepeatInterval)
            self.minimumDeleteRepeatInterval = max(0.01, minimumDeleteRepeatInterval)
            self.wordDeleteAfterRepeats = max(1, wordDeleteAfterRepeats)
        }
    }

    private struct KeyState: Sendable {
        let token: Int
        let literal: String
        let alternates: [String]
        let origin: InteractionPoint
        let allowsSwipe: Bool
        var isSwiping = false
        var alternatesVisible = false
        var alternateIndex = 0
    }

    private struct SpaceState: Sendable {
        let token: Int
        let origin: InteractionPoint
        var cursorMode = false
        var emittedCursorSteps = 0
    }

    private struct DeleteState: Sendable {
        let token: Int
        var repeatCount = 0
    }

    private enum ActiveState: Sendable {
        case key(KeyState)
        case space(SpaceState)
        case delete(DeleteState)
    }

    private let configuration: Configuration
    private var nextToken = 0
    private var active: ActiveState?

    public init(configuration: Configuration = Configuration()) {
        self.configuration = configuration
    }

    public mutating func handle(
        _ input: KeyboardInteractionInput
    ) -> [KeyboardInteractionEffect] {
        switch input {
        case let .press(target, point, time):
            return begin(target: target, at: point, time: time)
        case let .move(point, _):
            return move(to: point)
        case let .release(point, _):
            return release(at: point)
        case let .deadline(deadline):
            return fire(deadline)
        case .cancel:
            return cancel()
        }
    }

    private mutating func begin(
        target: KeyboardInteractionTarget,
        at point: InteractionPoint,
        time: TimeInterval
    ) -> [KeyboardInteractionEffect] {
        nextToken &+= 1
        let token = nextToken

        switch target {
        case let .key(literal, alternates, allowsSwipe):
            active = .key(
                KeyState(
                    token: token,
                    literal: literal,
                    alternates: alternates,
                    origin: point,
                    allowsSwipe: allowsSwipe
                )
            )
            var effects: [KeyboardInteractionEffect] = [
                .pressed(target),
                .preview(literal),
            ]
            if !alternates.isEmpty {
                effects.append(
                    .schedule(
                        KeyboardInteractionDeadline(
                            token: token,
                            dueTime: time + configuration.longPressDelay,
                            kind: .longPress
                        )
                    )
                )
            }
            effects.append(.feedback(.key))
            return effects

        case .space:
            active = .space(SpaceState(token: token, origin: point))
            return [
                .pressed(.space),
                .schedule(
                    KeyboardInteractionDeadline(
                        token: token,
                        dueTime: time + configuration.cursorActivationDelay,
                        kind: .cursorActivation
                    )
                ),
                .feedback(.key),
            ]

        case .delete:
            active = .delete(DeleteState(token: token))
            return [
                .pressed(.delete),
                .deleteBackward,
                .feedback(.key),
                .schedule(
                    KeyboardInteractionDeadline(
                        token: token,
                        dueTime: time + configuration.deleteRepeatDelay,
                        kind: .deleteRepeat
                    )
                ),
            ]
        }
    }

    private mutating func move(to point: InteractionPoint) -> [KeyboardInteractionEffect] {
        switch active {
        case var .key(state):
            if state.alternatesVisible {
                let rawIndex = Int(
                    ((point.x - state.origin.x) / configuration.alternateStep).rounded()
                )
                let index = min(state.alternates.count - 1, max(0, rawIndex))
                guard index != state.alternateIndex else { return [] }
                state.alternateIndex = index
                active = .key(state)
                return [
                    .showAlternates(state.alternates, selectedIndex: index),
                ]
            }

            if state.isSwiping {
                return [.swipeMoved(point)]
            }

            guard state.allowsSwipe,
                  state.origin.distance(to: point) >= configuration.swipeDistance else {
                return []
            }
            state.isSwiping = true
            active = .key(state)
            return [
                .preview(nil),
                .swipeBegan(state.origin),
                .swipeMoved(point),
            ]

        case var .space(state):
            let horizontalDistance = point.x - state.origin.x
            guard state.cursorMode
                    || abs(horizontalDistance) >= configuration.cursorActivationDistance else {
                return []
            }

            var effects: [KeyboardInteractionEffect] = []
            if !state.cursorMode {
                state.cursorMode = true
                effects += [.preview(nil), .feedback(.selection)]
            }

            let totalSteps = Int(horizontalDistance / configuration.cursorStep)
            let delta = totalSteps - state.emittedCursorSteps
            if delta != 0 {
                state.emittedCursorSteps = totalSteps
                effects.append(.moveCursor(delta))
            }
            active = .space(state)
            return effects

        case .delete, nil:
            return []
        }
    }

    private mutating func release(
        at point: InteractionPoint
    ) -> [KeyboardInteractionEffect] {
        let released = active
        active = nil

        switch released {
        case let .key(state):
            if state.isSwiping {
                return [.swipeEnded(point), .pressed(nil)]
            }
            if state.alternatesVisible {
                return [
                    .commitText(state.alternates[state.alternateIndex]),
                    .hideAlternates,
                    .preview(nil),
                    .pressed(nil),
                ]
            }
            return [
                .commitText(state.literal),
                .preview(nil),
                .pressed(nil),
            ]

        case let .space(state):
            if state.cursorMode {
                return [.pressed(nil)]
            }
            return [.commitText(" "), .pressed(nil)]

        case .delete:
            return [.pressed(nil)]

        case nil:
            return []
        }
    }

    private mutating func fire(
        _ deadline: KeyboardInteractionDeadline
    ) -> [KeyboardInteractionEffect] {
        switch (deadline.kind, active) {
        case let (.longPress, .key(state))
            where deadline.token == state.token
                && !state.isSwiping
                && !state.alternatesVisible
                && !state.alternates.isEmpty:
            var updated = state
            updated.alternatesVisible = true
            updated.alternateIndex = 0
            active = .key(updated)
            return [
                .showAlternates(updated.alternates, selectedIndex: 0),
                .feedback(.selection),
            ]

        case let (.cursorActivation, .space(state))
            where deadline.token == state.token && !state.cursorMode:
            var updated = state
            updated.cursorMode = true
            active = .space(updated)
            return [.preview(nil), .feedback(.selection)]

        case let (.deleteRepeat, .delete(state)) where deadline.token == state.token:
            var updated = state
            updated.repeatCount += 1
            active = .delete(updated)

            let acceleration = min(
                configuration.deleteRepeatInterval
                    - configuration.minimumDeleteRepeatInterval,
                Double(updated.repeatCount) * 0.01
            )
            let interval = configuration.deleteRepeatInterval - acceleration
            return [
                .deleteBackward,
                .feedback(.key),
                .schedule(
                    KeyboardInteractionDeadline(
                        token: updated.token,
                        dueTime: deadline.dueTime + interval,
                        kind: .deleteRepeat
                    )
                ),
            ]

        default:
            return []
        }
    }

    private mutating func cancel() -> [KeyboardInteractionEffect] {
        guard let active else { return [] }
        self.active = nil
        switch active {
        case let .key(state) where state.alternatesVisible:
            return [.hideAlternates, .preview(nil), .pressed(nil)]
        case .key:
            return [.preview(nil), .pressed(nil)]
        case .space, .delete:
            return [.pressed(nil)]
        }
    }
}
